core_data_modules/traced_data/io.py

```diff
@@ -232,28 +232,40 @@ def _make_empty_file():
         return io.StringIO("[]")
 
     @staticmethod
-    def _dataset_lut_from_messages_file(f):
+    def _dataset_lut_from_messages_file(f, schemes_to_parse):
         """
         Creates a lookup table of MessageID -> SchemeID -> Labels from the given Coda 2 messages file.
+        
+        Labels from schemes which are duplicates are automatically placed in the look-up table for the primary scheme.
 
         :param f: Coda 2 messages file.
         :type f: file-like
+        :param schemes_to_parse: Primary schemes to include in the returned LUT.
+        :type schemes_to_parse: iterable of Scheme
         :return: Lookup table.
-        :rtype: dict of str -> (dict of str -> list of dict)
+        :rtype: dict of str -> (dict of str -> list of Label)
         """
-        coda_dataset = dict()  # of MessageID -> (dict of SchemeID -> list of Label)
-
-        for msg in json.load(f):
-            schemes = dict()  # of SchemeID -> list of Label
-            coda_dataset[msg["MessageID"]] = schemes
-            msg["Labels"].reverse()  # Coda outputs most-recent first but easier to handle in Python if most-recent last
-            for label in msg["Labels"]:
-                scheme_id = label["SchemeID"]
-                if scheme_id not in schemes:
-                    schemes[scheme_id] = []
-                schemes[scheme_id].append(label)
-
-        return coda_dataset
+        coda_messages_file = json.load(f)
+        messages = []
+        for json_msg in coda_messages_file:
+            messages.append(Message.from_firebase_map(json_msg))
+
+        dataset_lut = dict()  # of MessageID -> (dict of SchemeID -> list of Label)
+        for msg in messages:
+            schemes_lut = dict()  # of SchemeID -> list of Label (for this message)
+            dataset_lut[msg.message_id] = schemes_lut
+
+            labels = list(msg.labels)
+            labels.reverse()  # Coda outputs most-recent first but easier to handle in Python if most-recent last
```

Comment:
I don't think we should do this, it means that when writing code you against this ordering you always need to think about whether you're using 'Coda ordering' or 'pipeline ordering'.

In Firestore we needed to do it this way round to minimise IO costs. 

I'm guessing that the main reason for this is because of the tracedata merging? I think I would rather put the complexity there rather than be passing around a data structure with it the other way round?

At any rate, let's discuss this, apologies I didn't catch it the first time I reviewed this.